browsergym/experiments/src/browsergym/experiments/loop.py

```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
                 self._summary_info = json.load(f)
         return self._summary_info
 
+    @property
+    def tape(self) -> dict:
+        """
+        TapeAgents (https://github.com/ServiceNow/TapeAgents) framework compatibility.
+        Exports experiment trace in the format of serialized tape.
+        Reuses tape segments if they were already placed in the agent_info during the experiment.
+
+        :returns: dict: serialized tape of the experiment
+        """
+        steps = []
+        for step_info in self.steps_info:
+            if "tape_segment" in step_info.agent_info["extra_info"]:
+                tape_segment = step_info.agent_info["extra_info"]["tape_segment"]
+            else:
+                tape_segment = self._create_tape_segment(step_info)
+            steps += tape_segment
+        metadata = dict(
+            id=str(uuid.uuid4()),
+            author="browsergym_agent",
+            result=self.get_exp_record(),
+        )
+        return dict(steps=steps, metadata=metadata)
+
+    def _create_tape_segment(self, step_info: StepInfo) -> list[dict]:
+        tape_segment = []
+        # extract observation step
+        if step_info.obs is not None:
+            screenshot: str = ""
+            screenshot_som: str = ""
+            obs_dict = step_info.obs.copy()
+            if "screenshot" in obs_dict:
+                screenshot = str(self.exp_dir / f"screenshot_step_{step_info.step}.png")
+                obs_dict.pop("screenshot")
+            if "screenshot_som" in obs_dict:
+                screenshot_som = str(self.exp_dir / f"screenshot_som_step_{step_info.step}.png")
+                obs_dict.pop("screenshot_som")
+            tape_segment.append(
+                dict(
+                    kind="browsergym_observation",
+                    metadata=dict(step=step_info.step),
+                    obs=obs_dict,
+                    screenshot=screenshot,
+                    screenshot_som=screenshot_som,
+                )
+            )
+
+        # extract thought step
+        think = step_info.agent_info.get("think", "")
+        if think:
+            tape_segment.append(
+                dict(kind="browsergym_thought", metadata={"step": step_info.step}, text=think)
+            )
+
+        # extract action step
+        action_str = step_info.action
+        # TODO: this is a naive way to split the arguments, make proper parsing later
```

Comment:
We have a parser in browsergym to extract function calls from raw text actions, you could just reuse it.

There is one caveat though, actions in browsergym can be arbitrary strings (python code, function calls, json). Even with our `function call` action set, agents can be given the freedom to execute several function calls in the same action (e.g. `type("a15", "Hello") click("a16")`).

Can TapeAgent handle arbitrary actions? Do you need the `name` and `arguments` key?

Reply:
You'll find an example here:
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/tests/core/test_actions_highlevel.py#L102

Reply:
The parser itself can be imported from here
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/browsergym/core/src/browsergym/core/action/parsers.py#L85

Reply:
I've implemented it, please take a look when you have time.